proxy/src/stream.rs

```diff
@@ -128,20 +128,42 @@ impl<S: AsyncWrite + Unpin> PqStream<S> {
     /// Write the error message using [`Self::write_message`], then re-throw it.
     /// Allowing string literals is safe under the assumption they might not contain any runtime info.
     /// This method exists due to `&str` not implementing `Into<anyhow::Error>`.
+    /// If `ctx` is provided and has testodrome_id set, error messages will be prefixed according to error kind.
     pub async fn throw_error_str<T>(
         &mut self,
         msg: &'static str,
         error_kind: ErrorKind,
+        ctx: Option<&crate::context::RequestContext>,
     ) -> Result<T, ReportedError> {
+        let formatted_msg = if let Some(ctx) = ctx {
+            if ctx.get_testodrome_id().is_some() {
+                match error_kind {
+                    ErrorKind::RateLimit | ErrorKind::User => msg.to_string(),
+                    ErrorKind::Service
+                    | ErrorKind::ServiceRateLimit
+                    | ErrorKind::ClientDisconnect => {
+                        format!("[proxy]: {msg}")
+                    }
+                    _ => format!("[{}]: {msg}", error_kind.to_metric_label()),
+                }
```

Comment:
Should we perhaps return some json encoded object instead? That way it's easy for us to add new data from `ctx` later.